src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,505 @@
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let builder = CGlyphOutline {
+        n_points: todo!(),
+        points: todo!(),
+        n_contours: todo!(),
+        contours: todo!(),
+    };
+    // also no clue what to do here
+    // let my_ol = caller.data().outline_glyph(GlyphId(glyph as u16), builder); // ??
+
+    let Ok(()) = memory.write(caller, outline as usize, todo!("builder result goes here")) else {
+        return 0;
+    };
+
+    1
+}
+
+// fn face_copy_table(font: u32, tag: u32, blob: *mut Blob) -> bool;
+// Copies the binary data in the OpenType table referenced by tag into the supplied blob structure.
+fn face_copy_table(mut caller: Caller<'_, ShapingData>, _font: u32, tag: u32, blob: u32) -> u32 {
+    // So here to copy stuff INTO the module, I need to copy it into its heap
+    // I should not assume that there is an area that's not written to,
+    // so the most straightforward way to get "clean" memory is to grow it by one page,
+    // and allocate there.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+    let tag = tag.to_be_bytes(); // be or le?
+    let Some(table) = caller.data().font.raw_face().table(Tag::from_bytes(&tag)) else {
+        return 0;
+    };
+
+    let eom = memory.data_size(&caller);
+    let Ok(_) = memory.grow(&mut caller.as_context_mut(), 1) else {
+        return 0;
+    };
+
+    let my_blob = Blob {
+        length: table.len() as u32,
+        data: eom as u32,
+    };
+
+    let Ok(()) = memory.write(
+        caller.as_context_mut(),
+        blob as usize,
+        bytemuck::bytes_of(&my_blob),
+    ) else {
+        return 0;
+    };
+
+    let Ok(()) = memory.write(caller.as_context_mut(), eom, table) else {
+        return 0;
+    };
+
+    1
+}
+
+// fn buffer_copy_contents(buffer: u32, cbuffer: *mut CBufferContents) -> bool;
+// Retrieves the contents of the host shaping engine's buffer into the buffer_contents structure. This should typically be called at the beginning of shaping.
+fn buffer_copy_contents(mut caller: Caller<'_, ShapingData>, _buffer: u32, cbuffer: u32) -> u32 {
+    // see face_copy_table for why we're growing memory.
+    let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+    let eom = memory.data(&caller).len();
+    let Ok(_) = memory.grow(&mut caller.as_context_mut(), 1) else {
+        return 0;
+    };
+
+    // I need these two to be the same lifetime it seems
+    let (mem_data, store_data) = memory.data_and_store_mut(&mut caller);
+
+    let rb_buffer = &store_data.buffer;
+    let length = rb_buffer.len;
+
+    let pos_loc = eom + length * core::mem::size_of::<hb_glyph_info_t>();
+    let end_loc = pos_loc + length * core::mem::size_of::<GlyphPosition>();
+
+    // This _should_ work ..
+    mem_data[eom..pos_loc].copy_from_slice(bytemuck::cast_slice(&rb_buffer.info));
+    mem_data[pos_loc..end_loc].copy_from_slice(bytemuck::cast_slice(&rb_buffer.pos));
```

Comment:
I think this could panic if the memory didn't grow enough

Reply:
As far as I understand the spec, it grows by one "page", which is 65,536 bytes. That fits 1638 characters (both info and pos). Maybe I can add a check.

Reply:
Aren't there multiple infos and pos structs? So if the buffer grows larger it could exceed the page size?

Reply:
GlyphPosition is size 20. I just divided 65,538 by 40 (both info and pos) to get that number. 1600 characters/letters is like 200 words ish. I don't know if shaper input is usually this large or not.